code/modules/atmospherics/machinery/components/machines/crystallizer.dm

```diff
@@ -0,0 +1,341 @@
+	var/recipe_type = null
+	///Stores the total amount of moles needed for the current recipe
+	var/total_recipe_moles = 0
+
+/obj/machinery/atmospherics/components/binary/crystallizer/Initialize()
+	. = ..()
+	internal = new(2000)
+
+/obj/machinery/atmospherics/components/binary/crystallizer/attackby(obj/item/I, mob/user, params)
+	if(!on)
+		if(default_deconstruction_screwdriver(user, "[base_icon]-open", "[base_icon]-off", I))
+			return
+	if(default_change_direction_wrench(user, I))
+		return
+	if(default_deconstruction_crowbar(I))
+		return
+	return ..()
+
+/obj/machinery/atmospherics/components/binary/crystallizer/default_change_direction_wrench(mob/user, obj/item/I)
+	. = ..()
+	if(!.)
+		return FALSE
+	SetInitDirections()
+	var/obj/machinery/atmospherics/node1 = nodes[1]
+	var/obj/machinery/atmospherics/node2 = nodes[2]
+	if(node1)
+		if(src in node1.nodes) //Only if it's actually connected. On-pipe version would is one-sided.
+			node1.disconnect(src)
+		nodes[1] = null
+	if(node2)
+		if(src in node2.nodes) //Only if it's actually connected. On-pipe version would is one-sided.
+			node2.disconnect(src)
+		nodes[2] = null
+
+	if(parents[1])
+		nullifyPipenet(parents[1])
+	if(parents[2])
+		nullifyPipenet(parents[2])
+
+	atmosinit()
+	node1 = nodes[1]
+	if(node1)
+		node1.atmosinit()
+		node1.addMember(src)
+	node2 = nodes[2]
+	if(node2)
+		node2.atmosinit()
+		node2.addMember(src)
+	SSair.add_to_rebuild_queue(src)
+	return TRUE
+
+/obj/machinery/atmospherics/components/binary/crystallizer/update_icon()
+	. = ..()
+	cut_overlays()
+
+	if(panel_open)
+		icon_state = "[base_icon]-open"
+	else if(on && is_operational)
+		icon_state = "[base_icon]-on"
+	else
+		icon_state = "[base_icon]-off"
+
+	add_overlay(getpipeimage(icon, "pipe", dir, COLOR_LIME, piping_layer))
+	add_overlay(getpipeimage(icon, "pipe", turn(dir, 180), COLOR_MOSTLY_PURE_RED, piping_layer))
+
+/obj/machinery/atmospherics/components/binary/crystallizer/attackby_secondary(mob/user)
+	if(!can_interact(user))
+		return
+	on = !on
+	investigate_log("was turned [on ? "on" : "off"] by [key_name(user)]", INVESTIGATE_ATMOS)
+	update_icon()
+
+///Checks if the gases in the input are the ones needed by the recipe
+/obj/machinery/atmospherics/components/binary/crystallizer/proc/check_gas_requirements()
+	var/datum/gas_mixture/contents = airs[2]
+	var/list/recipe = GLOB.gas_recipe_meta[recipe_type]
+	for(var/gas_type in recipe[META_RECIPE_REQUIREMENTS])
+		if(!contents.gases[gas_type] || !contents.gases[gas_type][MOLES])
+			return FALSE
+	return TRUE
+
+///Checks if the reaction temperature is inside the range of temperature + a little deviation
+/obj/machinery/atmospherics/components/binary/crystallizer/proc/check_temp_requirements()
+	var/datum/gas_mixture/contents = airs[2]
+	var/list/recipe = GLOB.gas_recipe_meta[recipe_type]
+	if(contents.temperature >= (recipe[META_RECIPE_MIN_TEMP] * MIN_DEVIATION_RATE) && contents.temperature <= (recipe[META_RECIPE_MAX_TEMP] * MAX_DEVIATION_RATE))
```

Comment:
the () here aren't really needed, and I think they'll end up making things harder to read in future

Reply:
okok